src/mongo_odm/query_builder.hpp

```diff
@@ -207,139 +612,810 @@ class BooleanExpr {
+     *
+     * @param builder A basic BSON core builder.
+     * @param Whether to wrap this expression inside a document.
+     */
+    void append_to_bson(bsoncxx::builder::core &builder, bool wrap = false) const {
+        if (wrap) {
+            builder.open_document();
+        }
+
+        _expr.append_to_bson(builder, false);
+        builder.key_view("$isolated");
+        builder.append(1);
+
+        if (wrap) {
+            builder.close_document();
+        }
+    }
+
+    /**
+     * Converts this query to BSON, in the form {$isolated: 1, <underlying expression BSON>}
+     */
+    operator bsoncxx::document::view_or_value() const {
+        auto builder = bsoncxx::builder::core(false);
+        append_to_bson(builder);
+        return {builder.extract_document()};
+    }
 
-/* Operator overloads for creating and combining expressions */
+   private:
+    const Expr _expr;
+};
 
-/* Overload comparison operators for name-value pairs to create expressions.
- * Separate operators are defined for bool types to prevent confusing implicit casting to bool
- * for non-bool types.
+/**
+ * Represents an update operator that modifies a certain elements.
+ * This creates BSON expressions of the form "$op: {field: value}",
+ * where $op can be $set, $inc, etc.
+ *
+ * This class stores the value to be compared by reference.
+ * To store a temporary or computed value, use UpdateExprValue.
+ *
+ * @tparam NvpT The type of the name-value pair
+ * @tparam U    The type of the value being compared.
  */
-template <typename Base, typename T, typename U,
-          typename = typename std::enable_if_t<!std::is_same<T, bool>::value>>
-constexpr ComparisonExpr<Base, T> operator==(const Nvp<Base, T> &lhs, const U &rhs) {
+template <typename NvpT, typename U>
+class update_expr {
+   public:
+    using field_type = NvpT;
+    constexpr update_expr(const NvpT &nvp, const U &val, const char *op)
+        : _nvp(nvp), _val(val), _op(op) {
+    }
+
+    /**
+     * Appends this query to a BSON core builder as a key-value pair "$op: {field: value}"
+     * @param builder A basic BSON core builder.
+     * @param Whether to wrap this expression inside a document.
+     */
+    void append_to_bson(bsoncxx::builder::core &builder, bool wrap = false) const {
+        if (wrap) {
+            builder.open_document();
+        }
+        builder.key_view(_op);
+        builder.open_document();
+        std::string s;
+        builder.key_view(_nvp.append_name(s));
+        append_value_to_bson(_val, builder);
+        builder.close_document();
+        if (wrap) {
+            builder.close_document();
+        }
+    }
+
+    /**
+     * Creates a BSON value of the form "$op: {field: value}"
+     */
+    operator bsoncxx::document::view_or_value() const {
+        auto builder = bsoncxx::builder::core(false);
+        append_to_bson(builder);
+        return {builder.extract_document()};
+    }
+
+   private:
+    NvpT _nvp;
```

Comment:
should this be a mutable value?

I'm a little confused by the various ways we capture our Nvp